### PR TITLE
added hashed level + tests

### PR DESCRIPTION
**Pull request by bharath2438**
@hameerabbasi I have added all the functions of hashed level along with a base test case. As for the locate test case, I wasn't sure of the format of the loop, so I'll implement it after discussion although the locate function works.

### CODE REVIEW

**Comment by hameerabbasi** [comment]
Could you add some tests of mixing it with other levels?

**Comment by bharath2438** [comment]
> Could you add some tests of mixing it with other levels?

I've added four combinations of hashed level with other levels

File: include/xsparse/levels/hashed.hpp [hunk 1]
```diff
@@ -36,8 +36,47 @@ namespace xsparse
...
+                class iterator;
+                using value_type =
+                    typename std::pair<typename BaseTraits::IK, typename BaseTraits::PK>;
+                using difference_type = typename std::make_signed_t<typename BaseTraits::PK>;
+                using pointer =
+                    typename std::pair<typename BaseTraits::IK, typename BaseTraits::PK>*;
+                using reference = std::pair<typename BaseTraits::IK, typename BaseTraits::PK>;
+                using iterator_type = iterator;
+
+                class iterator : public xtl::xbidirectional_iterator_base2<iteration_helper>
```
Comment (bharath2438) [inline include/xsparse/levels/hashed.hpp:48]: @hameerabbasi this is the iterator class and for defining ++, -- and == operators, I'll need a reference to an iterator of `iterator_type`. Should I wrap the `iterator` class inside another class to get this reference through the iterator of the vector or how should I go about doing it

File: include/xsparse/levels/hashed.hpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                class iterator;
+                using value_type =
+                    typename std::pair<typename BaseTraits::IK, typename BaseTraits::PK>;
+                using difference_type = typename std::make_signed_t<typename BaseTraits::PK>;
+                using pointer =
+                    typename std::pair<typename BaseTraits::IK, typename BaseTraits::PK>*;
+                using reference = std::pair<typename BaseTraits::IK, typename BaseTraits::PK>;
+                using iterator_type = iterator;
+
+                class iterator : xtl::xbidirectional_iterator_base2<iteration_helper>
```
Comment (hameerabbasi) [inline include/xsparse/levels/hashed.hpp:49]:
```suggestion
                class iterator : public xtl::xbidirectional_iterator_base2<iteration_helper>
```

So we can access all methods in the base class.

File: include/xsparse/levels/hashed.hpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                        return *this;
+                    }
+                };
+
+                explicit inline iteration_helper(typename BaseTraits::Level const& level,
+                                                 typename BaseTraits::I const i,
+                                                 typename BaseTraits::PKM1 const pkm1) noexcept
+                    : m_level(level)
+                    , m_pkm1(pkm1)
+                    , m_i(i)
```
Comment (hameerabbasi) [inline include/xsparse/levels/hashed.hpp:92]: Only a reference to the map needed as a member variable.

File: include/xsparse/levels/hashed.hpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+                inline iterator end() const noexcept
+                {
+                    return iterator{ m_level.m_crd[m_pkm1].end() };
+                }
+            };
+
+            iteration_helper iter_helper(typename BaseTraits::I i, typename BaseTraits::PKM1 pkm1)
+            {
+                return iteration_helper{ *static_cast<typename BaseTraits::Level*>(this), i, pkm1 };
```
Comment (hameerabbasi) [inline include/xsparse/levels/hashed.hpp:109]: You can directly pass the relevant map here.

File: include/xsparse/levels/hashed.hpp [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                m_crd[pkm1][ik] = pk;
+            }
+
+            inline void insert_edges(typename BaseTraits::PKM1 pkm1) const noexcept
+            {
+                m_crd_size += m_crd[pkm1].size();
+            }
+
+        private:
+            PK m_crd_size;
```
Comment (hameerabbasi) [inline include/xsparse/levels/hashed.hpp:152]: This is also needed in the constructors.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
This looks good to me overall. Just some minor comments.

File: include/xsparse/levels/hashed.hpp (hunk 4 above)
Comment (bharath2438) [reply include/xsparse/levels/hashed.hpp:109]: Is it okay to remove `BaseTraits::I i` as a parameter as it is not used anywhere?

File: include/xsparse/levels/hashed.hpp (hunk 5 above)
Comment (bharath2438) [reply include/xsparse/levels/hashed.hpp:152]: Is the size of type IK also needed along with `m_crd_size`?

File: include/xsparse/levels/hashed.hpp (hunk 4 above)
Comment (hameerabbasi) [reply include/xsparse/levels/hashed.hpp:109]: Yes, remove everything but the map.

File: include/xsparse/levels/hashed.hpp (hunk 5 above)
Comment (hameerabbasi) [reply include/xsparse/levels/hashed.hpp:152]: No, that isn't needed.

Comment (bharath2438) [reply include/xsparse/levels/hashed.hpp:148]: Done.

File: include/xsparse/levels/hashed.hpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            class iteration_helper
+            {
+                static_assert(std::is_nothrow_invocable_r_v<std::optional<typename BaseTraits::PK>,
+                                                            decltype(&BaseTraits::Level::locate),
+                                                            typename BaseTraits::Level&,
+                                                            typename BaseTraits::PKM1,
+                                                            typename BaseTraits::IK>);
+
+            private:
+                std::unordered_map<typename BaseTraits::IK, typename BaseTraits::PK> m_map;
```
Comment (hameerabbasi) [inline include/xsparse/levels/hashed.hpp:35]:
```suggestion
                std::unordered_map<typename BaseTraits::IK, typename BaseTraits::PK>& m_map;
```

File: include/xsparse/levels/hashed.hpp (hunk 5 above)
Comment (hameerabbasi) [reply include/xsparse/levels/hashed.hpp:148]: Oh, my bad, I misunderstood your question, yes, size is needed.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
I misunderstood you question about size, yes we do need it.

**Comment by bharath2438** [comment]
> I misunderstood you question about size, yes we do need it.

Done.

File: test/source/hashed_test.cpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+    std::vector<std::unordered_map<uintptr_t, uintptr_t>> const crd0{ umap1 };
+
+    std::vector<std::unordered_map<uintptr_t, uintptr_t>> const crd1{ umap2, umap3, umap4 };
+
+    xsparse::levels::hashed<std::tuple<>,
+                            uintptr_t,
+                            uintptr_t,
+                            std::vector<std::unordered_map<uintptr_t, uintptr_t>>>
+        h{ SIZE0, SIZE1, crd0 };
```
Comment (hameerabbasi) [inline test/source/hashed_test.cpp:66]: This will be the maximum value of `PK` for this level, plus `1`.

File: test/source/hashed_test.cpp [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    constexpr uintptr_t SIZE1 = 1;
+    constexpr uint8_t ZERO = 0;
+    constexpr uintptr_t SIZE2 = 3;
+    constexpr uintptr_t SIZE3 = 3;
+
+    std::unordered_map<uintptr_t, uintptr_t> const umap1{ { 0, 1 }, { 2, 0 }, { 1, 2 } };
+
+    std::unordered_map<uintptr_t, uintptr_t> const umap2{ { 5, 2 }, { 6, 1 }, { 4, 0 } };
+    std::unordered_map<uintptr_t, uintptr_t> const umap3{ { 2, 5 } };
+    std::unordered_map<uintptr_t, uintptr_t> const umap4{ { 3, 3 }, { 1, 4 }, { 0, 6 } };
```
Comment (hameerabbasi) [inline test/source/hashed_test.cpp:56]: The values of `PK` must be non-overlapping always, like they are here, and every value in the range `0..size_crd` must be present. `0` is included, `size_crd` isn't.

Comment (hameerabbasi) [reply test/source/hashed_test.cpp:56]: Same for all the other tests.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
Some comments on the tests.

File: test/source/hashed_test.cpp (hunk 8 above)
Comment (bharath2438) [reply test/source/hashed_test.cpp:56]: Okay, so should the `PK` values go from 0 to 3?

Comment (bharath2438) [reply test/source/hashed_test.cpp:56]: I've fixed the `size_crd` for all the test cases.

**Review by hameerabbasi — APPROVED** [review]
Waiting on CI for merge.

**Comment by hameerabbasi** [comment]
@bharath2438 Seems CI didn’t trigger, can you push an empty commit to force it?

**Comment by bharath2438** [comment]
> @bharath2438 Seems CI didn’t trigger, can you push an empty commit to force it?

yeah sure, will do

**Comment by hameerabbasi** [comment]
Thanks, @bharath2438!

**Comment by bharath2438** [comment]
Thanks a lot! Any reason why only the style check is happening?

**Comment by hameerabbasi** [comment]
See #17.

**Pull request merged**